cmd/controller-manager/app/controller-manager.go

```diff
@@ -107,12 +109,28 @@ func Run(opts *options.Options, stopChan <-chan struct{}) error {
 		panic(err)
 	}
 
-	setOptionsByKubeFedConfig(opts)
-
 	if err := utilfeature.DefaultFeatureGate.SetFromMap(opts.FeatureGates); err != nil {
 		klog.Fatalf("Invalid Feature Gate: %v", err)
 	}
 
+	runningInCluster := len(masterURL) == 0 && len(kubeconfig) == 0
+	if runningInCluster && len(opts.Config.KubeFedNamespace) == 0 {
+		// for in-cluster deployment set the namespace associated
+		// with the service account token
+		data, err := ioutil.ReadFile("/var/run/secrets/kubernetes.io/serviceaccount/namespace")
+		if err != nil {
+			klog.Fatalf("An error occurred while attempting to discover the kubefed namespace from the service account: %v", err)
+		}
+		opts.Config.KubeFedNamespace = strings.TrimSpace(string(data))
+	}
+
+	setOptionsByKubeFedConfig(opts)
+
+	// validate if a kubefed namespace is configured
+	if len(opts.Config.KubeFedNamespace) == 0 {
```

Comment:
This check needs to occur before `setOptionsByKubeFedConfig` is called to ensure that the check's error message will appear in the logs instead of a complaint that a namespace must be set from an attempt to load a KubeFedConfig resource.

Reply:
Fixed